File: pythonSolutions/day2_2.py
with open("pythonSolutions\day2.txt") as file:
    lines = file.readlines()

sum_of_powers = 0

for line in lines:
    num_of_maxred = 0
    num_of_maxblue = 0
    num_of_maxgreen = 0
    
    game, cubes_drawn = line.split(":")
    game_ID = int(game.split(" ")[1])
    sets = cubes_drawn.split(";")

    for s in sets: # one set of a line
        cubes = s.split(",")

        for cube in cubes:  
            cube = cube.strip()
            number, color = cube.split(" ")
            number = int(number)

            if (color == "red" and number > num_of_maxred):
                num_of_maxred = number
            elif (color == "green" and number > num_of_maxgreen):
                num_of_maxgreen = number
            elif (color == "blue" and number > num_of_maxblue):
                num_of_maxblue = number

    power_of_set = num_of_maxblue * num_of_maxgreen * num_of_maxred
    sum_of_powers += power_of_set
        
print(sum_of_powers)

File: AdventOfCode2023/Day2_2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCode2023
{
    public static class Day2_2
    {
        public static void Execute()
        {
            String line;
            try
            {
                StreamReader sr = new StreamReader("../Day2.txt");

                line = sr.ReadLine();

                List<string> games = new List<string>();

                while (line != null)
                {
                    games.Add(line);
                    line = sr.ReadLine();
                }
                sr.Close();

                int sumOfPowers = 0;

                //# go through each line (separated by line)
                //# go through each allCubesDrawn (separated by ;)
                //# go through each cube (separated by ,)
                //#if out of possible range - allCubesDrawn the flag to false
                //#if the flag is still true (aka the game is possible) - add its ID to the sum

                for (int i = 0; i < games.Count; i++)
                {
                    int numberOfMaxRed = 0;
                    int numberOfMaxGreen = 0;
                    int numberOfMaxBlue = 0;
                    
                    string allCubesDrawn = games[i].Split(":", StringSplitOptions.RemoveEmptyEntries)[1];
                    string[] sets = allCubesDrawn.Split(";", StringSplitOptions.RemoveEmptyEntries);
                    foreach (string set in sets)
                    {
                        string[] cubes = set.Trim().Split(",", StringSplitOptions.RemoveEmptyEntries);
                        foreach (string cube in cubes)
                        {
                            int number = Convert.ToInt32(cube.Trim().Split(" ", StringSplitOptions.RemoveEmptyEntries)[0]);
                            string color = cube.Trim().Split(" ", StringSplitOptions.RemoveEmptyEntries)[1];
                            if (color == "red" && number > numberOfMaxRed)
                            {
                                numberOfMaxRed = number;
                            }
                            if (color == "green" && number > numberOfMaxGreen)
                            {
                                numberOfMaxGreen = number;
                            }
                            if (color == "blue" && number > numberOfMaxBlue)
                            {
                                numberOfMaxBlue = number;
                            }
                        }
                    }
                    int powerOfSet = numberOfMaxRed * numberOfMaxGreen * numberOfMaxBlue;
                    sumOfPowers += powerOfSet;
                }
                Console.WriteLine("Sum of powers: " + sumOfPowers);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception: " + e.Message);
            }
            finally
            {
                Console.WriteLine("Executing finally block.");
            }
        }
    }
}
